### ISSUE TITLE
SiconosVector implementation

### ISSUE THREAD

**Issue by bremond** [siconos/siconos#127]
SiconosVector encapsulates a pointer (DenseVect*, SparseVect*) to some structure that manages a dynamic storage, with itself manages internaly another pointer to allocated data. A boolean is also associated to distinguish between dense and sparse case.

SiconosVector objects are used nearly everywhere in Siconos and there
is three main problems with this implementation:

 - it is really not memory efficient for small, fixed size, vectors (two step heap allocation + useless boolean padded as an int + useless size stored internally somewhere). 

 - it is heap based. Any temporary storage requested in a loop needs an allocation. Even if we change

   SP::SiconosVector temp

   by
   
   SiconosVector temp

   This hardly change something (only the first pointer and the
   boolean go on the stack and at least two allocations are still performed).

 - it does not allow easy extension to other type of storage, such
   as fixed, stack based vectors.

As SiconosVector is ubiquitous and any change to it is very sensible,
we may think about two big steps in order to improve the situation:

 1. keep the Dense and Sparse storage functionality and keep
    SiconosVector interface as it is. Remove the useless boolean
    thanks to visitable mechanisms, remove the union pointer and
    install a pointer on an abstract storage class. The key point is
    that derivations of the storage class are visitable: Dense
    storage, Sparse storage, Bounded storage.  Dense and sparse
    remains dynamically allocated. A BoundedVector inherits from
    SiconosVector and makes the storage pointer to point on its own
    bounded storage which is not heap allocated.  At this level, we
    gain a bounded storage at the memory cost of a remaining pointer
    in the structure.

 2. remove the Dense and sparse storage functionality in order to
    remove the storage pointer.  Make two distinct vectors :
    Densevector SparseVector. This would need a change in the api, as
    SiconosVector constructors would disappear (they could be changed
    by some factory function). At this level, we gain a bounded
    storage at no memory cost.

The first step is implemented in this fork

https://github.com/bremond/siconos/tree/kernel-new-siconosvector

With a whole change in SiconosVector implementation, it provides some
new bounded SiconosVector : SiconosVector3 SiconosVector4
SiconosVector6 SiconosVector7 

There are not used at the moment.

tests are mainly ok, BouncingBall is ok, cube.py is ok.

Just a quick test, to show the allocation slowdown with SiconosVector:

  SiconosVector v;
  v.resize(7);
  v.fill(0.);

  for (unsigned int i=0; i<10000000; ++i)
  {
    SiconosVector u = v;
    v(i%7) = (double) i + (u)(6 - i%7);
  }

real	0m2.219s
user	0m2.216s
sys	0m0.000s

is nearly 10 time slower than

  SiconosVector7 v;
  v.resize(7);
  v.fill(0.);

  for (unsigned int i=0; i<10000000; ++i)
  {
    SiconosVector7 u = v;
    v(i%7) = (double) i + (u)(6 - i%7);
  }

real	0m0.243s
user	0m0.240s
sys	0m0.000s

Some points missing

  - no time or memory comparison with master SiconosVector

  - no tests on bounded vectors,

  - it uses bounded storage (a useless size remains) but ublas does not
    implement fixed size vectors. 

  - the visitor mechanisms is a bit complicate and should be cleaned and documented
    (the apply_visitor function is meant to be used at other places)

  - the python interface is not done. Note that it may be possible to
    have bi-directional interface (a change to some bounded vector in
    python could be visible in C++)

  - usage remains to be done : in Newton Euler, Collision Detection, Bullet, Occ, ...

**Comment by radarsat1** [comment]
Nice initiative and very promising results!  I have been thinking it would even be nice to have a class that "knows" a little more about the NE case of SiconosVector7 being divided into position and quaternion parts, SiconosVector7 might be a good basis for that.

Anyways, I will take a look more in-depth next week.  Regards.

**Comment by bremond** [comment]
Yes siconosVector7 is probably useless and is not the correct
representation of what is needed, it is really better to consider the
position as a specific type which is an association of a translation
(SiconosVector3) and a rotation (a boost quaternion).

For bounded vectors we need SiconosVector3 for translation and
SiconosVector6 for velocity.

**Comment by radarsat1** [comment]
An initial look at this, I like the `apply_visitor` and `make_visitor` functions, I find they clarify the use of visitors rather well.  My main problem with visitors is that it separates the definition of operations from where they are applied, which imho makes it hard to read the code.  However good naming conventions help immensely with this.  I particularly like the choice of subscript "Op" for operations on vectors.  I wonder if some of the further readability issues can be alleviated by clever use of C++11/14 features, but that is for the (far) future I guess.

I am unable to completely compile, some undefined reference linker errors deep in the template code which is a bit hard to debug. e.g.,

    ../../../libsiconos_kernel.so.4.1.0: undefined reference to `void OpAxpby::operator()<boost::numeric::ublas....

**Comment by radarsat1** [comment]
Another thing I note, if we are going to add lots of classes with very generic names like Zero, Size, Fill, etc, we _must_ start using C++ namespaces, otherwise it will be impossible to use Siconos as a library in a larger program.

**Comment by radarsat1** [comment]
Regarding the linker errors, for example there is a function:

    template<>
    void Zero::operator()<DenseVectStorage> (DenseVectStorage& dv);

but it is not defined anywhere.  Is there maybe a `SiconosVectorOperators.cpp` that you forgot to add?

**Comment by bremond** [comment]
I have added the missing file SiconosVectorOperators.cpp
It contains some specializations.

I agree visitor mechanism is painfull to read, (it is also painfull to
write in the case of more than one dispatch (all the cases where there
is two storages to visit)).

With variadic templates, c++11 would lead really simpler code for the
VisitorMaker, but I do not see an alternative to this visitor builder as
virtual templated functions are not for tomorrow.

I don't know for the readability, maybe a trick with some lambda
functions that you would write where you call apply_visitor.
templated lambda are possible in c++14
http://stackoverflow.com/questions/3575901/can-lambda-functions-be-templated
So we can imagine an apply_visitor taking a generic visitor with
the templated lambda as argument. Everything in one line...
An emulation exists with boost::phoenix, I may have a try.

Another point with c++11 is rvalue references which seems interesting
for the return value of visitors, but I do not know if it is possible
to use them in this case. For c++03, the emulation with boost::move is
really too heavy.

**Comment by radarsat1** [comment]
I tried for a while this morning using C++14 lambdas actually, and came close to something that works, but it still thinks one of the parameters is SiconosVectorStorage. :(

I'll try again sometime soon.. ;P  it's fun, but it would really be nice to have this:

    apply_visitor([](const auto & y, auto & x){
        x.resize(y.size());
        x = y;
    }, storage(v), storage(w));

I got very close (I think), but not quite there unfortunately.  Metaprogramming :P 
What I did so far was to create a Visitor that takes a function as a parameter,

~~~
template<typename F, typename T, typename P>
void apply_visitor(F&& f, T& obj, P& param)
{
  FuncVisitor<F,P> visitor(param, f);
  obj.accept(visitor);
};

template<typename F, typename P>
struct FuncVisitor : public ParamVisitor<P>
{
  F&& _f;

  virtual ~FuncVisitor() {};

  typedef ParamVisitor<P> base_type;
  FuncVisitor(typename base_type::arguments_type& st, F&& f)
                       : base_type(st), _f(f) {};

  template<typename T>
  void visit (const T& storage)
  {
    _f(this->param(), storage.internal_data);
  }

  template<typename T>
  void visit (T& storage)
  {
    _f(this->param(), storage.internal_data);
  }
};
~~~

which didn't work until I added some specialisation for DenseVectStorage:

~~~
void visit (DenseVectStorage& storage)
{
  _f(storage.internal_data, this->param());
}
~~~

At this point it gets as far as evaluating `_f` but unfortunately I then get the error,
~~~
error: ‘class SiconosVectorStorage’ has no member named ‘resize’
     x.resize(y.size());
~~~
So it fails to sufficiently specialize the "templated" lambda.  I also tried adapting `make_visitor` to take an extra function parameter but I couldn't get this to compile, with some error related to `Experimental::Call` being "recursively required".  At this point I stopped.

**Comment by bremond** [comment]
The copy operator as it is implemented here leads to a triple dispatch
problem for the compiler and this leads to a double dispatch for us (*),
you should first try with a simple dispatch one, the simplest one is the
zero operator, then you have fill, resize operators with simple
arguments.

For the double dispatch this is where I need the InterBiOp which build an
internal visitor to visit the second argument. This should be also possible
with lambda in one line but it may be a bit more complicated...

Also, some general cascading mechanism maybe interesting here instead of
InternBiOp because even more dispatching is needed in matrices operators.

(*) I don't know if I am correct, what I call a n+1 level dispatch
    problem at the level of the compiler is:

 object.function(x1, x2, ..., xn)

equivalent to

 function(object, x1, x2, ..., xn)

where object, x1, x2 ... xn are all instances of some classes that
inherits from other classes and only the base classes are known at
compile time.

The first dispatch on 'object' type is done by the compiler thanks to
the vtable. Then as there is no multi-dispatch in c++, for more
dispatching we have to use visitor patterns for the n remaining
dispatch.

So for object.function(x1)  a double dispatch for the compiler that
leads to a simple dispatch for us, this is a simple call to a
visitor :

void objectclass::function(X& x1)
{
  apply_visitor<FunctionVisitor>(x1);
}

where x1 is an instance of a class X1 which derives from class X and you
want to know X1.

This is equivalent to:

void objectclass::function(X& x1)
{
  FunctionVisitor visitor;
  x1.accept(visitor);
}

For a triple dispatch object.function(x1, x2) at the level of the
compiler, you have to visit both arguments at the same time:

void objectclass::function(X& x1, X& x2)
{
  apply_visitor<FunctionVisitor>(x1, x2);
}

this visits x1 then create a visitor with the correct type of x1 as
argument which then visit the argument x2 and then apply the operator on
the correct type of x1 and x2.

This seems strange to have a 2 dispatch for copy, but here we are
visiting a storage inside the object. This is because I wanted to keep
SiconosVector as it is without changing the whole siconos API, I don't
know if there is a better solution.

**Comment by radarsat1** [comment]
Thanks, I didn't fully understand the need for InternBiOp.  I think your last comment should go into developer_notes.org ;)

**Comment by xhub** [comment]
This PR looks exciting. Any updates?

**Comment by bremond** [comment]
Not for the moment, what is missing is matrix vector operations. This is not a huge work, but I am on occ bindings now. So I close the PR (it was just an advertisement).

We had an interesting discussion with @radarsat1. It appears that with templated lambda, this would change drastically this kind of code, but it is not until c++14.

**Comment by vacary** [comment]
Transferred in wiki

https://github.com/siconos/siconos/wiki/Wish-list-and-enhancement
